git-keeper.py

```diff
@@ -94,8 +99,15 @@ def main():
         default="",
         help="Path of [comma delimited] subfolder[s]" " in bucket to store backups",
     )
+    parser.add_argument(
+        "--commit",
+        action="store_true",
+        help="use SHA of last commit instead of date as sub-subfolder",
+    )
     args = parser.parse_args()
     subfolders = [str(subfolder) for subfolder in args.subfolders.split(",")]
+    global commit
```

Comment:
can we avoid global vars by calculating `sub_subfolder` outside of `git_clone_upload`?

Also if we want to easily sort s3 objects by key to find latest backups, we can combine date and sha together.

Reply:
we are looping through list of repos, so we have different `sub_subfolder` for each repo

Reply:
surely, can use as addition parameter to function call instead of `global`, if main thing is avoiding using it

Reply:
yes, it is always better to use globals. You need to be able to pass either date or sha depending on the and not rely on globals which are always complicated to follow.

Reply:
no need for additional parameter, as you already renamed `date` to `sub_subfolder` for `git_clone_upload`, which indicate `git_clone_upload` no longer care the value of it, can just use it as passed in.

can calculate it like this:


```python
for repo in repolist:
    if args.commit:
        sha = git("--git-dir=" + repo_dir, "rev-parse", "--verify", "HEAD")
        sub_subfolder = f"{date}-{sha}" # or just sha if no need to sort
    else:
        sub_subfolder = date
    ...
```

Reply:
we need to clone repo first, to be able to get latest commit hash.
If we switch to obtaining latest commit hash from remote repo - it's possible that it changes between remote call and clone, so we can have inconsistency

Reply:
oops, then let's pass it as an additional param to `git_clone_upload`, shouldn't use `global` to pass info to a function. You can keep the old name `date` in function parameter, and use `commit` as a flag to decide what to use as `sub_subfolder`, maybe worth creating a small helper make it clear:

```python
def _sub_subfolder(repo_dir, date, commit):
    if not commit:
        return date
    sha = git("--git-dir=" + repo_dir, "rev-parse", "--verify", "HEAD") 
    return f"{date}-{sha}" # or just sha if no need to sort
```

Reply:
+1 to removing global (for some reason my yesterday's comment about it disappeared 😢 )

Reply:
updated, does it looks OK now?

Reply:
please don't mutate passed in parameter `sub_subfolder = sha`, keep the old `date` in param, and conditional construct `sub_subfolder` or extract out a function.

```python
    sub_subfolder = sh.git("--git-dir=" + repo_dir, "rev-parse", "--verify", "HEAD") if commit else date
```

just wonder without `date` in s3 path, how to quickly know which one is the lasted backup?

Reply:
backups with commit hash in path supposed to be stored in additional folder near regular daily backups.
It's implied that folders with commits_id contains every commit and lifecycle should clean old backups from that folder more aggressively, like every 25 hours or so.
But realized S3 have no ability to sort on modification date, you need to download all metadata at least and sort locally (or lambda). I'll rethink folder structure for that.

Reply:
yes, append commit sha to current date string should solve the overwrite issue and keep the sort feature `/{date}-{sha}/`